### DISCUSSION TITLE
Add suboption to Hidden (HD) for fading hitobjects in while the player is struggling.

### DISCUSSION THREAD

**Discussion by DavidBeh** [ppy/osu#30871] [Ideas]
I would like to add an (non-default) option to the osu!std Hidden mod for adjusting to the difficulty the player has with the current part of the beatmap. Either by fading hitobjects back in as soon as the player STARTS STRUGGLING until his GAMEPLAY STABALIZES; or by increasing the fadeout duration and/or target fading opacity based on the PLAYERS PERFORMANCE.

This would allow the player to more easily get back to stable gameplay when having lost his “flow”. For example, For example, when I hit a hitobjects at the wrong beat, this often leads to a domino effect of misses, resulting in failure, especially for streams.

The Implementation could roughly be structured in the following way:

* Calculating the things printed in CAPS in the first paragraph. Please let me know if you have any suggestions on how to determine these.
* Hiding and Revealing the Hitobjects, maybe in a universal way for all rulesets (by calling `ApplyIncreasedVisibilityState` for example) based on the calculations above. Here I would also like to hear your suggestions on how to react to the calculation results (opacity/fade duration).

I have already created a prototype which sets the approach circles opacity based on the users combo (Currently: Opacity = Median(0, combo/25, 1) * 100%). This is somewhat similar to https://github.com/ppy/osu/discussions/18114#discussioncomment-3281984
Demo of my implementation: https://www.youtube.com/watch?v=R89rHRldPPo
Source Code: https://github.com/DavidBeh/osu/tree/hidden-diffadjusting

**Comment by cloudrac3r** [comment]
> Calculating the things printed in CAPS in the first paragraph. Please let me know if you have any suggestions on how to determine these.

The way this is done for other mods like Muted and No Scope is to have normal gameplay at 0 combo and as the combo increases the mods start to take full effect, or in the case of Flashlight have a more significant impact (smaller radius) with higher combo. That would probably work well for Hidden too.

**Comment by DavidBeh** [comment]
As suggested by @cloudrac3r I will continue with making it combo-based. I plan to expose the following members to implementers of `ModHidden`:
* overridable method for supplying combo delta value of a hitobject when judged (to ignore ticks for example)
* the opacity to apply to hitobjects
* Maybe the time at which the last combo break was

I would then make ModHidden handle the easing (non linear!), keeping track of combo, etc,.
The things I have described are already partially implemented. Current state: https://www.youtube.com/watch?v=R_NH8qfJu4k
The frame drops are from OBS.

**Reply by cloudrac3r** [reply]
That looks really good!
